handlers/lookup_handler.go

```diff
@@ -13,56 +15,63 @@ import (
 // This handler assumes that the Logger already has the ctrl_gk field set.
 type LookupHandler struct {
 	Logger    *zap.Logger
-	WorkQueue WorkQueueProducer
+	WorkQueue ctrl.WorkQueueProducer
+	Gvk       schema.GroupVersionKind
 
 	Lookup func(runtime.Object) ([]runtime.Object, error)
 }
 
-func (e *LookupHandler) enqueueMapped(obj meta_v1.Object, addUpdateDelete string) {
-	logger := e.loggerForObj(obj)
+func (e *LookupHandler) enqueueMapped(logger *zap.Logger, obj meta_v1.Object) {
+	logger = e.loggerForObj(logger, obj)
 	objs, err := e.Lookup(obj.(runtime.Object))
 	if err != nil {
 		logger.Error("Failed to lookup objects", zap.Error(err))
 		return
 	}
+	if len(objs) == 0 {
+		logger.Debug("Lookup function returned zero results")
+	}
 	for _, o := range objs {
 		metaobj := o.(meta_v1.Object)
-		logger.With(logz.Controller(metaobj)).Sugar().
-			Infof("Enqueuing controller object because controlled object was %s", addUpdateDelete)
-		e.WorkQueue.Add(QueueKey{
+		logger.Sugar().Infof("Enqueuing looked up object '%s/%s'", obj.GetNamespace(), obj.GetName())
+		e.WorkQueue.Add(ctrl.QueueKey{
 			Namespace: metaobj.GetNamespace(),
 			Name:      metaobj.GetName(),
 		})
 	}
 }
 
 func (e *LookupHandler) OnAdd(obj interface{}) {
-	e.enqueueMapped(obj.(meta_v1.Object), "added")
+	logger := e.Logger.With(logz.Operation(ctrl.AddedOperation))
+	e.enqueueMapped(logger, obj.(meta_v1.Object))
 }
 
 func (e *LookupHandler) OnUpdate(oldObj, newObj interface{}) {
-	e.enqueueMapped(newObj.(meta_v1.Object), "updated")
+	logger := e.Logger.With(logz.Operation(ctrl.UpdatedOperation))
+	e.enqueueMapped(logger, newObj.(meta_v1.Object))
 }
 
 func (e *LookupHandler) OnDelete(obj interface{}) {
 	metaObj, ok := obj.(meta_v1.Object)
+	logger := e.Logger.With(logz.Operation(ctrl.DeletedOperation))
 	if !ok {
 		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
 		if !ok {
-			e.Logger.Sugar().Errorf("Delete event with unrecognized object type: %T", obj)
+			logger.Sugar().Errorf("Delete event with unrecognized object type: %T", obj)
 			return
 		}
 		metaObj, ok = tombstone.Obj.(meta_v1.Object)
 		if !ok {
-			e.Logger.Sugar().Errorf("Delete tombstone with unrecognized object type: %T", tombstone.Obj)
+			logger.Sugar().Errorf("Delete tombstone with unrecognized object type: %T", tombstone.Obj)
 			return
 		}
 	}
-	e.enqueueMapped(metaObj, "deleted")
+	e.enqueueMapped(logger, metaObj)
 }
 
 // loggerForObj returns a logger with fields for a controlled object.
-func (e *LookupHandler) loggerForObj(obj meta_v1.Object) *zap.Logger {
-	return e.Logger.With(logz.Namespace(obj), logz.Object(obj),
-		logz.ObjectGk(obj.(runtime.Object).GetObjectKind().GroupVersionKind().GroupKind()))
+func (e *LookupHandler) loggerForObj(logger *zap.Logger, obj meta_v1.Object) *zap.Logger {
+	return logger.With(logz.Namespace(obj),
+		logz.Object(obj),
+		logz.ObjectGk(e.Gvk.GroupKind()))
```

Comment:
Same - object_gk should be set already.

Reply:
We have a field called "controller_gk" which is always present on all operations associated with that controller, starting from construction of the constroller. These include logs that don't involve processing an object.

"object_gk" is not set by `Generic`. It sets `controller_gk` to identify which controller a log message is associated with, even during the periods where it is not processing an object. For example, this distintion is useful in the LookupHandler because, for example:

1. `controller_gk` could be `ServiceDescriptor`
2. `object_gk` could be `RoleBinding`.

I agree that this distinction is less useful in the `Process` method of the generic_worker because `controller_gk` is always equal to `object_gk`. However, we were using `logz.ControllerName`, which was... confusingly named. To make this consistent with how the rest of the code behaves, I thought it would be reasonable to include both `object_name` and `object_gk` for *any* processing that happens within ctrl.

I renamed `controller_name` to `delegate_name` to be better.

There was another change I did in this PR which was to fix the original `logz.ObjectGk` call, because Kubernetes strips the APIVersion and Kind from objects coming from Watch/List. This means the ResourceEventHandlers need to be passed in the Gvk that they should log.

In summary the changes I made were:

1. Get rid of `controller_name`, because the term is overloaded. This was truly being used in a random unpredictable way. I've replaced it with a `delegate_name`
2. Fix `object_gk` because it was always empty string
3. Made usage of `object_gk` consistent.
4. Cleaned up logging of operation

How do we use this in structured logging? Let's say I want to figure out *everything* that has been constructed or processed for a particular controller. I can query splunk for `controller_gk`. If I wanted to look at only processing I can query for `object_gk`. If I want to find all relevant logs I can query for `object_gk` and `delegate_gk`, which should give me all object updates, as well as the logs of anything which was queued as a result of a lookup.

Reply:
Ok, thanks for reminding me how it all works - I have forgotten =) You are right, it all makes sense. Please remove/amend those misleading comments on handler/etc structs to reflect what is really expected of the injected logger. Thanks!